### PR TITLE
#1554 Add space to empty anchor tags to fix internal links.

### PR DESCRIPTION
**Pull request by rpedela**
Internal links (anchor tags) are not linked correctly when the anchor's content is empty. This adds a space with a height of 0 when it is empty so that internal links are linked correctly.

@partychen deserves credit for this fix. I am just making the PR.

### CODE REVIEW

File: src/lib/pdfconverter.cc [hunk 1]
```diff
@@ -546,13 +546,26 @@ void PdfConverterPrivate::findLinks(QWebFrame * frame, QVector<QPair<QWebElement
...
 		QString h=elm.attribute("href");
-		if (h.isEmpty()) h=elm.attribute("ns0:href");
+		if (h.isEmpty()) {
+			h=elm.attribute("ns0:href");
+
+			//Internal links (anchor tags) are not linked correctly
+			//when the anchor's content is empty. This adds a space with
+			//a height of 0 when it is empty so that internal links
+			//are linked correctly.
+			if(elm.firstChild().isNull()){
```
Comment (ashkulz) [inline src/lib/pdfconverter.cc]: This should be outside i.e.
```c
QString h=elm.attribute("href");
if (h.isEmpty()) h=elm.attribute("ns0:href");

// Internal links ...
if (elm.firstChild().isNull()) {
```
Otherwise, this will get executed only when there is no `href` attribute but only a `ns0:href` attribute -- unless that is what is wanted?

File: src/lib/pdfconverter.cc [hunk 2]
```diff
@@ -546,13 +546,26 @@ void PdfConverterPrivate::findLinks(QWebFrame * frame, QVector<QPair<QWebElement
 		uexternal  = PageObject::webPageToObject[frame->page()]->settings.useExternalLinks;
 	}
 	if (!ulocal && !uexternal) return;
-	foreach (const QWebElement & elm, frame->findAllElements("a")) {
+	foreach (QWebElement elm, frame->findAllElements("a")) {
```
Comment (ashkulz) [inline src/lib/pdfconverter.cc]: Why is this required?

**Comment by ashkulz** [comment]
Also, please rework the commit message (see previous commits) so that it includes the comments you have put in the PR as a part of the commit message. Additionally, I'd prefer to keep the comments out of the code (`git blame` will tell you why the change was made).

You can use `git commit --author "Name <Email@Domain>" to override the author, with you as the committer. Please make sure you rebase/amend the commit so that there is only a single commit which needs to be merged.

File: src/lib/pdfconverter.cc (hunk 1 above)
Comment (rpedela) [inline src/lib/pdfconverter.cc]: The logic is that if the ```a``` tag has an empty ```href``` attribute then it must be an anchor tag, but it only adds the space if there aren't any children. I tested the following scenarios which all worked:

1. ```<a name="abc"></a>```
2. ```<a id="abc"></a>```
3. ```<a name="abc" href=""></a>```
4. ```<a id="abc" href=""></a>```

If I move the code outside then it will execute for all ```a``` tags and will replace the link text with a space. I tested that just to make sure.

File: src/lib/pdfconverter.cc (hunk 2 above)
Comment (rpedela) [inline src/lib/pdfconverter.cc]: We need to call ```set()``` functions to inject ```nbsp;``` so ```const``` needs to be removed. The non-const reference gives me the following compiler error:

```
g++ -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT4_STATICPLUGIN_TEXTCODECS -DVERSION=0.12.3.2 -DFULL_VERSION=0.12.3.2 -DBUILDING_DLL -DQT_NODLL -DQT_NO_DEBUG -DQT_WEBKIT_LIB -DQT_SVG_LIB -DQT_XMLPATTERNS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I/home/rpedela/wkhtmltopdf/qt/mkspecs/linux-g++ -I/home/rpedela/wkhtmltopdf/src/lib -I../../../qt/include/QtCore -I../../../qt/include/QtNetwork -I../../../qt/include/QtGui -I../../../qt/include/QtXmlPatterns -I../../../qt/include/QtSvg -I../../../qt/include/QtWebKit -I../../../qt/include -I/home/rpedela/wkhtmltopdf/include -I. -I/home/rpedela/wkhtmltopdf/src/lib -I. -o pdfconverter.o /home/rpedela/wkhtmltopdf/src/lib/pdfconverter.cc
In file included from ../../../qt/include/QtCore/qglobal.h:1:0,
                 from ../../../qt/include/QtCore/../../../../../qt/src/corelib/global/qnamespace.h:45,
                 from ../../../qt/include/QtCore/qnamespace.h:1,
                 from ../../../qt/include/QtCore/../../../../../qt/src/corelib/kernel/qobjectdefs.h:45,
                 from ../../../qt/include/QtCore/qobjectdefs.h:1,
                 from ../../../qt/include/QtCore/../../../../../qt/src/corelib/kernel/qobject.h:47,
                 from ../../../qt/include/QtCore/qobject.h:1,
                 from ../../../qt/include/QtCore/QObject:1,
                 from /home/rpedela/wkhtmltopdf/src/lib/converter.hh:24,
                 from /home/rpedela/wkhtmltopdf/src/lib/converter_p.hh:24,
                 from /home/rpedela/wkhtmltopdf/src/lib/pdfconverter_p.hh:24,
                 from /home/rpedela/wkhtmltopdf/src/lib/pdfconverter.cc:22:
/home/rpedela/wkhtmltopdf/src/lib/pdfconverter.cc: In member function 'void wkhtmltopdf::PdfConverterPrivate::findLinks(QWebFrame*, QVector<QPair<QWebElement, QString> >&, QVector<QPair<QWebElement, QString> >&, QHash<QString, QWebElement>&)':
../../../qt/include/QtCore/../../../../../qt/src/corelib/global/qglobal.h:2500:34: error: invalid initialization of non-const reference of type 'QWebElement&' from an rvalue of type 'const QWebElement'
     for (variable = *_container_.i;; __extension__ ({--_container_.brk; break;}))
                                  ^
../../../qt/include/QtCore/../../../../../qt/src/corelib/global/qglobal.h:2564:21: note: in expansion of macro 'Q_FOREACH'
 #    define foreach Q_FOREACH
                     ^
/home/rpedela/wkhtmltopdf/src/lib/pdfconverter.cc:549:2: note: in expansion of macro 'foreach'
  foreach (QWebElement & elm, frame->findAllElements("a")) {
  ^
make[1]: *** [pdfconverter.o] Error 1
```

Do you have any suggestions?

File: src/lib/pdfconverter.cc (hunk 1 above)
Comment (ashkulz) [inline src/lib/pdfconverter.cc]: Okay, so then this approach is wrong/incomplete. The `ns0:href` is generated by the TOC support, so it is quite possible that it won't work there. A better approach is to check directly for the condition that you want (`<a name="not_empty"></a>`) rather than the absence of something (missing `href` on `<a>`) and work on that directly. Additionally, you might want to check for `id` tags on headers/divs and add some content there too (they can also be used as the target of an href).

File: src/lib/pdfconverter.cc (hunk 2 above)
Comment (ashkulz) [inline src/lib/pdfconverter.cc]: Yeah, I forgot about the `setStyleProperty` calls below -- that's OK :+1:

File: src/lib/pdfconverter.cc (hunk 1 above)
Comment (rpedela) [inline src/lib/pdfconverter.cc]: The check could be changed to also check that the ```a``` tag has an ```id``` or ```name``` too. That would be a stronger check and less likely to produce side effects. However an empty ```href``` attribute is the key condition.

For ```ns0:href```, are you saying that we also need to check that ```ns0:href``` is empty?

Comment (ashkulz) [inline src/lib/pdfconverter.cc]: Yes, the TOC is generated via XSLT which unfortunately adds tags with the `ns0:` namespace, which is why you see code for that being specifically handled.

Comment (rpedela) [inline src/lib/pdfconverter.cc]: You make a good point about other tags being used as targets which I didn't think about. Here are the cases I can think of:

1. An ```a``` tag with either ```id``` or ```name``` attributes, empty ```href``` or ```ns0:href``` attributes, and no children.
2. Any other anchor tags (e.g. ```div, p, h1```) that have an ```id``` and no children. I can look up the full list of tags than can be used as an anchor.

Am I missing any cases?

Comment (ashkulz) [inline src/lib/pdfconverter.cc]: Looks OK to me. Also, just noticed -- please target `master` and not the `0.12.3.x` branch.

**Comment by ashkulz** [comment]
@rpedela: any update?

**Comment by rpedela** [comment]
Nope. I haven't had time to work on it. Hopefully soon.

**Comment by rpedela** [comment]
I have tested ```<div id="something"></div>``` and ```<div name="something"></div>``` which worked without any code changes, and based on my understanding of the code it should work for any other HTML tag with an ```id```. I think it is just ```<a id="something"></a>``` and ```<a name="something"></a>``` which is broken.

> Also, just noticed -- please target master and not the 0.12.3.x branch.

How about both ```master``` and ```0.12.3.x```? This is pretty major bug in my opinion, and I would prefer to not use ```master``` in production. Of course, I could use my own branch until the next stable release, but I am sure I am not the only one who needs this bug fixed in production.

**Comment by ashkulz** [comment]
I'm planning to release `0.12.4` pretty soon, so it will be in the next stable release. I keep `0.12.x.x` only for critical security updates in bundled libraries e.g. OpenSSL.
